File: 07_后端/lingyi_service/app/schemas/subcontract.py
"""Pydantic schemas for subcontract module (TASK-002)."""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Generic
from typing import List
from typing import Optional
from typing import TypeVar

from pydantic import BaseModel
from pydantic import Field
from pydantic import model_validator

T = TypeVar("T")


class ApiResponse(BaseModel, Generic[T]):
    """Unified API response envelope."""

    code: str
    message: str
    data: T


class SubcontractCreateRequest(BaseModel):
    """Request payload of creating subcontract order."""

    supplier: str = Field(..., min_length=1, max_length=140)
    item_code: str = Field(..., min_length=1, max_length=140)
    company: Optional[str] = Field(default=None, max_length=140)
    bom_id: int = Field(..., ge=1)
    planned_qty: Decimal = Field(..., gt=0)
    process_name: str = Field(..., min_length=1, max_length=100)


class SubcontractCreateData(BaseModel):
    """Create subcontract order response data."""

    name: str
    company: str


class SubcontractListQuery(BaseModel):
    """List query parameters."""

    supplier: Optional[str] = None
    status: Optional[str] = None
    from_date: Optional[str] = None
    to_date: Optional[str] = None
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=200)


class SubcontractListItem(BaseModel):
    """Subcontract order list row."""

    id: int
    subcontract_no: str
    supplier: str
    item_code: str
    company: Optional[str] = None
    bom_id: int
    process_name: str
    planned_qty: Decimal
    subcontract_rate: Decimal = Decimal("0")
    issued_qty: Decimal = Decimal("0")
    received_qty: Decimal = Decimal("0")
    inspected_qty: Decimal = Decimal("0")
    rejected_qty: Decimal = Decimal("0")
    accepted_qty: Decimal = Decimal("0")
    gross_amount: Decimal = Decimal("0")
    deduction_amount: Decimal = Decimal("0")
    net_amount: Decimal = Decimal("0")
    status: str
    resource_scope_status: str = "ready"
    latest_issue_outbox_id: Optional[int] = None
    latest_issue_sync_status: Optional[str] = None
    latest_issue_stock_entry_name: Optional[str] = None
    latest_issue_idempotency_key: Optional[str] = None
    latest_issue_error_code: Optional[str] = None
    latest_receipt_outbox_id: Optional[int] = None
    latest_receipt_sync_status: Optional[str] = None
    latest_receipt_stock_entry_name: Optional[str] = None
    latest_receipt_idempotency_key: Optional[str] = None
    latest_receipt_error_code: Optional[str] = None
    created_at: datetime


class SubcontractListData(BaseModel):
    """List result payload."""

    items: List[SubcontractListItem]
    total: int
    page: int
    page_size: int


class SubcontractSettlementSummary(BaseModel):
    """Settlement summary for candidates/preview/lock responses."""

    line_count: int
    total_qty: Decimal
    gross_amount: Decimal
    deduction_amount: Decimal
    net_amount: Decimal


class SubcontractSettlementCandidateItem(BaseModel):
    """Settlement candidate row based on inspection fact."""

    inspection_id: int
    settlement_line_key: Optional[str] = None
    subcontract_id: int
    subcontract_no: str
    company: str
    supplier: str
    item_code: str
    process_name: str
    receipt_batch_no: str
    inspected_at: Optional[datetime] = None
    inspected_by: Optional[str] = None
    inspected_qty: Decimal
    accepted_qty: Decimal
    rejected_qty: Decimal
    rejected_rate: Decimal
    subcontract_rate: Decimal
    gross_amount: Decimal
    deduction_amount: Decimal
    net_amount: Decimal
    settlement_status: str
    statement_id: Optional[int] = None
    statement_no: Optional[str] = None


class SubcontractSettlementCandidatesData(BaseModel):
    """Settlement candidate list payload."""

    items: List[SubcontractSettlementCandidateItem]
    total: int
    page: int
    page_size: int
    summary: SubcontractSettlementSummary


class SubcontractSettlementPreviewRequest(BaseModel):
    """Settlement preview request."""

    inspection_ids: Optional[List[int]] = None
    company: Optional[str] = Field(default=None, max_length=140)
    supplier: Optional[str] = Field(default=None, max_length=140)
    from_date: Optional[str] = None
    to_date: Optional[str] = None
    item_code: Optional[str] = Field(default=None, max_length=140)
    process_name: Optional[str] = Field(default=None, max_length=100)


class SubcontractSettlementPreviewData(BaseModel):
    """Settlement preview response."""

    company: Optional[str] = None
    supplier: Optional[str] = None
    line_count: int
    total_qty: Decimal
    gross_amount: Decimal
    deduction_amount: Decimal
    net_amount: Decimal
    items: List[SubcontractSettlementCandidateItem]


class SubcontractSettlementLockRequest(BaseModel):
    """Settlement lock request."""

    statement_id: Optional[int] = Field(default=None, ge=1)
    statement_no: Optional[str] = Field(default=None, max_length=64)
    inspection_ids: List[int] = Field(default_factory=list)
    idempotency_key: str = Field(..., min_length=1, max_length=128)
    remark: Optional[str] = Field(default=None, max_length=200)


class SubcontractSettlementLockData(BaseModel):
    """Settlement lock response."""

    operation_id: int
    idempotency_key: str
    idempotent_replay: bool
    locked_count: int
    gross_amount: Decimal
    deduction_amount: Decimal
    net_amount: Decimal
    locked_items: List[SubcontractSettlementCandidateItem]


class SubcontractSettlementReleaseRequest(BaseModel):
    """Settlement lock release request."""

    statement_id: Optional[int] = Field(default=None, ge=1)
    statement_no: Optional[str] = Field(default=None, max_length=64)
    inspection_ids: List[int] = Field(default_factory=list)
    idempotency_key: str = Field(..., min_length=1, max_length=128)
    reason: str = Field(..., min_length=1, max_length=200)


class SubcontractSettlementReleaseData(BaseModel):
    """Settlement lock release response."""

    operation_id: int
    idempotency_key: str
    idempotent_replay: bool
    released_count: int
    released_items: List[SubcontractSettlementCandidateItem]


class SubcontractReceiptDetailItem(BaseModel):
    """Receipt detail row in subcontract detail response."""

    receipt_batch_no: str
    receipt_warehouse: Optional[str] = None
    item_code: Optional[str] = None
    color: Optional[str] = None
    size: Optional[str] = None
    batch_no: Optional[str] = None
    uom: Optional[str] = None
    received_qty: Decimal
    sync_status: str
    stock_entry_name: Optional[str] = None
    inspect_status: Optional[str] = None
    idempotency_key: Optional[str] = None
    received_by: Optional[str] = None
    received_at: Optional[datetime] = None


class SubcontractDetailData(BaseModel):
    """Subcontract detail payload."""

    id: int
    subcontract_no: str
    supplier: str
    item_code: str
    company: Optional[str] = None
    bom_id: int
    process_name: str
    planned_qty: Decimal
    subcontract_rate: Decimal = Decimal("0")
    issued_qty: Decimal = Decimal("0")
    received_qty: Decimal = Decimal("0")
    inspected_qty: Decimal = Decimal("0")
    rejected_qty: Decimal = Decimal("0")
    accepted_qty: Decimal = Decimal("0")
    gross_amount: Decimal = Decimal("0")
    deduction_amount: Decimal = Decimal("0")
    net_amount: Decimal = Decimal("0")
    status: str
    settlement_status: Optional[str] = None
    resource_scope_status: str
    scope_error_code: Optional[str] = None
    latest_issue_outbox_id: Optional[int] = None
    latest_issue_sync_status: Optional[str] = None
    latest_issue_stock_entry_name: Optional[str] = None
    latest_issue_idempotency_key: Optional[str] = None
    latest_receipt_outbox_id: Optional[int] = None
    latest_receipt_sync_status: Optional[str] = None
    latest_receipt_stock_entry_name: Optional[str] = None
    latest_receipt_idempotency_key: Optional[str] = None
    receipts: List[SubcontractReceiptDetailItem] = Field(default_factory=list)
    inspections: List["SubcontractInspectionDetailItem"] = Field(default_factory=list)
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class IssueMaterialItem(BaseModel):
    """Material issue line input."""

    material_item_code: str = Field(..., min_length=1, max_length=140)
    required_qty: Decimal = Field(..., gt=0)
    issued_qty: Decimal = Field(..., gt=0)

    @model_validator(mode="before")
    @classmethod
    def _compat_item_alias(cls, value):
        if isinstance(value, dict) and "material_item_code" not in value:
            item_code = value.get("item_code")
            if isinstance(item_code, str):
                value["material_item_code"] = item_code
        return value


class IssueMaterialRequest(BaseModel):
    """Issue material request."""

    idempotency_key: str = Field(..., min_length=1, max_length=128)
    warehouse: str = Field(..., min_length=1, max_length=140)
    materials: List[IssueMaterialItem] | None = None

    @model_validator(mode="before")
    @classmethod
    def _compat_materials_alias(cls, value):
        if isinstance(value, dict) and "materials" not in value and "items" in value:
            value["materials"] = value.get("items")
        return value


class IssueMaterialData(BaseModel):
    """Issue material response payload."""

    outbox_id: int
    issue_batch_no: str
    sync_status: str
    stock_entry_name: Optional[str] = None


class ReceiveRequest(BaseModel):
    """Receive subcontract goods request."""

    idempotency_key: str = Field(..., min_length=1, max_length=128)
    receipt_warehouse: str = Field(..., min_length=1, max_length=140)
    received_qty: Decimal = Field(..., gt=0)
    item_code: Optional[str] = Field(default=None, max_length=140)
    color: Optional[str] = Field(default=None, max_length=64)
    size: Optional[str] = Field(default=None, max_length=64)
    batch_no: Optional[str] = Field(default=None, max_length=140)
    uom: Optional[str] = Field(default=None, max_length=32)


class ReceiveData(BaseModel):
    """Receive response data."""

    outbox_id: int
    receipt_batch_no: str
    sync_status: str
    stock_entry_name: Optional[str] = None


class InspectRequest(BaseModel):
    """Inspect subcontract goods request."""

    receipt_batch_no: str = Field(..., min_length=1, max_length=64)
    idempotency_key: str = Field(..., min_length=1, max_length=128)
    inspected_qty: Decimal = Field(..., gt=0)
    rejected_qty: Decimal = Field(..., ge=0)
    deduction_amount_per_piece: Decimal = Field(default=Decimal("0"), ge=0)
    remark: Optional[str] = Field(default=None, max_length=200)


class InspectData(BaseModel):
    """Inspect response data."""

    inspection_no: str
    receipt_batch_no: str
    inspected_qty: Decimal
    accepted_qty: Decimal
    rejected_qty: Decimal
    rejected_rate: Decimal
    gross_amount: Decimal
    deduction_amount: Decimal
    net_amount: Decimal
    status: str


class SubcontractInspectionDetailItem(BaseModel):
    """Inspection detail row in subcontract detail response."""

    inspection_no: str
    receipt_batch_no: str
    inspected_qty: Decimal
    accepted_qty: Decimal
    rejected_qty: Decimal
    rejected_rate: Decimal
    subcontract_rate: Decimal
    gross_amount: Decimal
    deduction_amount_per_piece: Decimal
    deduction_amount: Decimal
    net_amount: Decimal
    inspected_by: Optional[str] = None
    inspected_at: Optional[datetime] = None
    remark: Optional[str] = None


class SubcontractStockSyncRetryRequest(BaseModel):
    """Manual stock sync retry target selector."""

    outbox_id: int = Field(..., ge=1)
    stock_action: str = Field(..., min_length=1, max_length=32)
    idempotency_key: str = Field(..., min_length=1, max_length=128)
    reason: Optional[str] = Field(default=None, max_length=200)


class SubcontractStockSyncRetryData(BaseModel):
    """Manual stock sync retry response."""

    outbox_id: int
    stock_action: str
    status: str
    next_retry_at: Optional[datetime] = None


class SubcontractStockSyncRunOnceData(BaseModel):
    """Internal stock sync worker run result."""

    dry_run: bool
    batch_size: int
    would_process_count: int
    processed_count: int
    succeeded_count: int
    failed_count: int
    dead_count: int

File: 06_前端/lingyi-pc/src/api/subcontract.ts
import { request, type ApiResponse } from '@/api/request'

export type NumericLike = string | number
export type SubcontractWriteOperation =
  | 'create'
  | 'issue_material'
  | 'receive'
  | 'inspect'
  | 'settlement_preview'
  | 'settlement_lock'
  | 'release'

const SUBCONTRACT_SCENARIO_PATTERN = /^Z003-SUBCONTRACT-\d{8}-\d{3}$/
const SUBCONTRACT_SCENARIO_EXTRACT_PATTERN = /Z003-SUBCONTRACT-\d{8}-\d{3}/
const SUBCONTRACT_REQUEST_ID_PATTERN = /^[A-Za-z0-9_.-]{1,64}$/

export interface SubcontractOrderPayload {
  supplier: string
  item_code: string
  bom_id: number
  planned_qty: number
  process_name: string
}

export interface SubcontractOrderListItem {
  id: number
  subcontract_no: string
  supplier: string
  item_code: string
  company?: string | null
  bom_id: number
  process_name: string
  planned_qty: string
  subcontract_rate: string
  issued_qty: string
  received_qty: string
  inspected_qty: string
  rejected_qty: string
  accepted_qty: string
  gross_amount: string
  deduction_amount: string
  net_amount: string
  status: string
  resource_scope_status: string
  latest_issue_outbox_id?: number | null
  latest_issue_sync_status?: string | null
  latest_issue_stock_entry_name?: string | null
  latest_issue_idempotency_key?: string | null
  latest_issue_error_code?: string | null
  latest_receipt_outbox_id?: number | null
  latest_receipt_sync_status?: string | null
  latest_receipt_stock_entry_name?: string | null
  latest_receipt_idempotency_key?: string | null
  latest_receipt_error_code?: string | null
  production_plan_id?: number | null
  work_order?: string | null
  created_at: string
}

export interface SubcontractOrderListData {
  items: SubcontractOrderListItem[]
  total: number
  page: number
  page_size: number
}

export interface SubcontractWriteCarrierPayload {
  request_id: string
  idempotency_key: string
  scenario_tag: string
  source_ref: string
  subcontract_ref: string
  supplier_ref: string
  work_order_ref: string
  operation: SubcontractWriteOperation
  item_code: string
  quantity: NumericLike
  status_action: string
}

export interface SubcontractIssueMaterialItemInput {
  material_item_code: string
  required_qty: NumericLike
  issued_qty: NumericLike
}

export interface SubcontractCreateRequestPayload extends SubcontractWriteCarrierPayload {
  operation: 'create'
  supplier: string
  item_code: string
  company?: string | null
  bom_id: number
  planned_qty: NumericLike
  process_name: string
  sales_order?: string | null
  sales_order_item?: string | null
  production_plan_id?: number | null
  work_order?: string | null
  job_card?: string | null
}

export interface SubcontractCreateResponseData {
  name: string
  company: string
}

export interface SubcontractIssueMaterialRequestPayload extends SubcontractWriteCarrierPayload {
  operation: 'issue_material'
  warehouse: string
  materials: SubcontractIssueMaterialItemInput[]
}

export interface SubcontractIssueMaterialResponseData {
  outbox_id: number
  issue_batch_no: string
  sync_status: string
  stock_entry_name?: string | null
}

export interface SubcontractReceiveRequestPayload extends SubcontractWriteCarrierPayload {
  operation: 'receive'
  receipt_warehouse: string
  received_qty: NumericLike
  item_code: string
  color?: string | null
  size?: string | null
  batch_no?: string | null
  uom?: string | null
}

export interface SubcontractReceiveResponseData {
  outbox_id: number
  receipt_batch_no: string
  sync_status: string
  stock_entry_name?: string | null
}

export interface SubcontractInspectRequestPayload extends SubcontractWriteCarrierPayload {
  operation: 'inspect'
  receipt_batch_no: string
  inspected_qty: NumericLike
  rejected_qty: NumericLike
  deduction_amount_per_piece?: NumericLike
  remark?: string | null
}

export interface SubcontractInspectResponseData {
  inspection_no: string
  receipt_batch_no: string
  inspected_qty: NumericLike
  accepted_qty: NumericLike
  rejected_qty: NumericLike
  rejected_rate: NumericLike
  gross_amount: NumericLike
  deduction_amount: NumericLike
  net_amount: NumericLike
  status: string
}

export interface SubcontractSettlementCandidateListItem {
  inspection_id: number
  settlement_line_key?: string | null
  subcontract_id: number
  subcontract_no: string
  company: string
  supplier: string
  item_code: string
  process_name: string
  receipt_batch_no: string
  inspected_at?: string | null
  inspected_by?: string | null
  inspected_qty: NumericLike
  accepted_qty: NumericLike
  rejected_qty: NumericLike
  rejected_rate: NumericLike
  subcontract_rate: NumericLike
  gross_amount: NumericLike
  deduction_amount: NumericLike
  net_amount: NumericLike
  settlement_status: string
  statement_id?: number | null
  statement_no?: string | null
}

export interface SubcontractSettlementSummary {
  line_count: number
  total_qty: NumericLike
  gross_amount: NumericLike
  deduction_amount: NumericLike
  net_amount: NumericLike
}

export interface SubcontractSettlementCandidatesData {
  items: SubcontractSettlementCandidateListItem[]
  total: number
  page: number
  page_size: number
  summary: SubcontractSettlementSummary
}

export interface SubcontractSettlementPreviewRequestPayload extends SubcontractWriteCarrierPayload {
  operation: 'settlement_preview'
  inspection_ids?: number[]
  company?: string | null
  supplier?: string | null
  from_date?: string | null
  to_date?: string | null
  filter_item_code?: string | null
  process_name?: string | null
}

export interface SubcontractSettlementPreviewData {
  company?: string | null
  supplier?: string | null
  line_count: number
  total_qty: NumericLike
  gross_amount: NumericLike
  deduction_amount: NumericLike
  net_amount: NumericLike
  items: SubcontractSettlementCandidateListItem[]
}

export interface SubcontractSettlementLockRequestPayload extends SubcontractWriteCarrierPayload {
  operation: 'settlement_lock'
  statement_id?: number | null
  statement_no?: string | null
  inspection_ids: number[]
  remark?: string | null
}

export interface SubcontractSettlementLockData {
  operation_id: number
  idempotency_key: string
  idempotent_replay: boolean
  locked_count: number
  gross_amount: NumericLike
  deduction_amount: NumericLike
  net_amount: NumericLike
  locked_items: SubcontractSettlementCandidateListItem[]
}

export interface SubcontractSettlementReleaseRequestPayload extends SubcontractWriteCarrierPayload {
  operation: 'release'
  statement_id?: number | null
  statement_no?: string | null
  inspection_ids: number[]
  reason: string
}

export interface SubcontractSettlementReleaseData {
  operation_id: number
  idempotency_key: string
  idempotent_replay: boolean
  released_count: number
  released_items: SubcontractSettlementCandidateListItem[]
}

export interface SubcontractReceiptDetailItem {
  receipt_batch_no: string
  receipt_warehouse?: string | null
  item_code?: string | null
  color?: string | null
  size?: string | null
  batch_no?: string | null
  uom?: string | null
  received_qty: string
  sync_status: string
  stock_entry_name?: string | null
  inspect_status?: string | null
  idempotency_key?: string | null
  received_by?: string | null
  received_at?: string | null
}

export interface SubcontractInspectionDetailItem {
  inspection_no: string
  receipt_batch_no: string
  inspected_qty: string
  accepted_qty: string
  rejected_qty: string
  rejected_rate: string
  subcontract_rate: string
  gross_amount: string
  deduction_amount_per_piece: string
  deduction_amount: string
  net_amount: string
  inspected_by?: string | null
  inspected_at?: string | null
  remark?: string | null
}

export interface SubcontractOrderDetailData extends SubcontractOrderListItem {
  scope_error_code?: string | null
  settlement_status?: string | null
  latest_issue_outbox_id?: number | null
  latest_issue_sync_status?: string | null
  latest_issue_stock_entry_name?: string | null
  latest_issue_idempotency_key?: string | null
  latest_receipt_outbox_id?: number | null
  latest_receipt_sync_status?: string | null
  latest_receipt_stock_entry_name?: string | null
  latest_receipt_idempotency_key?: string | null
  receipts: SubcontractReceiptDetailItem[]
  inspections: SubcontractInspectionDetailItem[]
  updated_at?: string | null
}

const toQuery = (params: Record<string, string | number | undefined>): string => {
  const query = new URLSearchParams()
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== '') {
      query.append(key, String(value))
    }
  })
  return query.toString()
}

const fnvCarrierCode = (value: string): string => {
  const normalized = value.trim()
  let hashValue = 2166136261
  for (let index = 0; index < normalized.length; index += 1) {
    hashValue ^= normalized.charCodeAt(index)
    hashValue = Math.imul(hashValue, 16777619) >>> 0
  }
  return hashValue.toString(16).toUpperCase().padStart(8, '0').slice(-3)
}

const subcontractOperationCode = (operation: SubcontractWriteOperation): 'CR' | 'IM' | 'RV' | 'IN' | 'SP' | 'SL' | 'RL' => {
  if (operation === 'create') return 'CR'
  if (operation === 'issue_material') return 'IM'
  if (operation === 'receive') return 'RV'
  if (operation === 'inspect') return 'IN'
  if (operation === 'settlement_preview') return 'SP'
  if (operation === 'settlement_lock') return 'SL'
  return 'RL'
}

export const buildSubcontractScenarioTag = (): string => {
  const now = new Date()
  const day = `${now.getFullYear()}${String(now.getMonth() + 1).padStart(2, '0')}${String(now.getDate()).padStart(2, '0')}`
  const seq = String(Math.floor(Math.random() * 1000)).padStart(3, '0')
  return `Z003-SUBCONTRACT-${day}-${seq}`
}

export const extractSubcontractScenarioTag = (value: string): string | null => {
  const matched = value.trim().match(SUBCONTRACT_SCENARIO_EXTRACT_PATTERN)
  return matched ? matched[0] : null
}

export const ensureSubcontractScenarioTag = (value: string): string => {
  const normalized = value.trim()
  if (SUBCONTRACT_SCENARIO_PATTERN.test(normalized)) {
    return normalized
  }
  return buildSubcontractScenarioTag()
}

export const buildSubcontractRequestId = ({
  scenarioTag,
  operation,
  idempotencyKey,
  sourceRef,
  subcontractRef,
  supplierRef,
  workOrderRef,
  itemCode,
  statusAction,
}: {
  scenarioTag: string
  operation: SubcontractWriteOperation
  idempotencyKey: string
  sourceRef: string
  subcontractRef: string
  supplierRef: string
  workOrderRef: string
  itemCode: string
  statusAction: string
}): string => {
  const normalizedScenarioTag = ensureSubcontractScenarioTag(scenarioTag)
  const requestId = `${normalizedScenarioTag}-SC-${subcontractOperationCode(operation)}-${fnvCarrierCode(idempotencyKey)}-${fnvCarrierCode(sourceRef)}-${fnvCarrierCode(subcontractRef)}-${fnvCarrierCode(supplierRef)}-${fnvCarrierCode(workOrderRef)}-${fnvCarrierCode(itemCode)}-${fnvCarrierCode(statusAction)}`
  if (!SUBCONTRACT_REQUEST_ID_PATTERN.test(requestId)) {
    throw new Error('request_id 编码非法')
  }
  return requestId
}

export const fetchSubcontractOrders = async (params: {
  supplier?: string
  status?: string
  from_date?: string
  to_date?: string
  page: number
  page_size: number
}): Promise<ApiResponse<SubcontractOrderListData>> => {
  const query = toQuery(params)
  return request<SubcontractOrderListData>(`/api/subcontract/?${query}`)
}

export const fetchSubcontractOrderDetail = async (orderId: number): Promise<ApiResponse<SubcontractOrderDetailData>> => {
  return request<SubcontractOrderDetailData>(`/api/subcontract/${orderId}`)
}

export const fetchSubcontractSettlementCandidates = async (params: {
  company?: string
  supplier?: string
  item_code?: string
  process_name?: string
  status?: string
  from_date?: string
  to_date?: string
  page?: number
  page_size?: number
} = {}): Promise<ApiResponse<SubcontractSettlementCandidatesData>> => {
  const query = toQuery({
    company: params.company,
    supplier: params.supplier,
    item_code: params.item_code,
    process_name: params.process_name,
    status: params.status,
    from_date: params.from_date,
    to_date: params.to_date,
    page: params.page ?? 1,
    page_size: params.page_size ?? 20,
  })
  return request<SubcontractSettlementCandidatesData>(`/api/subcontract/settlement-candidates${query ? `?${query}` : ''}`)
}

export const createSubcontractOrder = async (
  payload: SubcontractCreateRequestPayload,
): Promise<ApiResponse<SubcontractCreateResponseData>> => {
  return request<SubcontractCreateResponseData>('/api/subcontract/', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', 'X-Request-ID': payload.request_id },
    body: JSON.stringify(payload),
  })
}

export const issueSubcontractMaterial = async (
  orderId: number,
  payload: SubcontractIssueMaterialRequestPayload,
): Promise<ApiResponse<SubcontractIssueMaterialResponseData>> => {
  return request<SubcontractIssueMaterialResponseData>(`/api/subcontract/${orderId}/issue-material`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', 'X-Request-ID': payload.request_id },
    body: JSON.stringify(payload),
  })
}

export const receiveSubcontractOrder = async (
  orderId: number,
  payload: SubcontractReceiveRequestPayload,
): Promise<ApiResponse<SubcontractReceiveResponseData>> => {
  return request<SubcontractReceiveResponseData>(`/api/subcontract/${orderId}/receive`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', 'X-Request-ID': payload.request_id },
    body: JSON.stringify(payload),
  })
}

export const inspectSubcontractOrder = async (
  orderId: number,
  payload: SubcontractInspectRequestPayload,
): Promise<ApiResponse<SubcontractInspectResponseData>> => {
  return request<SubcontractInspectResponseData>(`/api/subcontract/${orderId}/inspect`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', 'X-Request-ID': payload.request_id },
    body: JSON.stringify(payload),
  })
}

export const previewSubcontractSettlement = async (
  payload: SubcontractSettlementPreviewRequestPayload,
): Promise<ApiResponse<SubcontractSettlementPreviewData>> => {
  return request<SubcontractSettlementPreviewData>('/api/subcontract/settlement-preview', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', 'X-Request-ID': payload.request_id },
    body: JSON.stringify(payload),
  })
}

export const lockSubcontractSettlement = async (
  payload: SubcontractSettlementLockRequestPayload,
): Promise<ApiResponse<SubcontractSettlementLockData>> => {
  return request<SubcontractSettlementLockData>('/api/subcontract/settlement-locks', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', 'X-Request-ID': payload.request_id },
    body: JSON.stringify(payload),
  })
}

export const releaseSubcontractSettlement = async (
  payload: SubcontractSettlementReleaseRequestPayload,
): Promise<ApiResponse<SubcontractSettlementReleaseData>> => {
  return request<SubcontractSettlementReleaseData>('/api/subcontract/settlement-locks/release', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', 'X-Request-ID': payload.request_id },
    body: JSON.stringify(payload),
  })
}
